pulsar-common/src/main/java/org/apache/pulsar/common/naming/NameCache.java

```diff
@@ -0,0 +1,156 @@
+ */
+abstract class NameCache<V> {
+    // Cache instances using ConcurrentHashMap and SoftReference to allow garbage collection to clear unreferenced
+    // entries when heap memory is running low.
+    private final ConcurrentMap<String, SoftReferenceValue> cache = new ConcurrentHashMap<>();
+    // Reference queue to hold cleared soft references, which will be used to remove entries from the cache.
+    private final ReferenceQueue<? super V> referenceQueue = new ReferenceQueue<>();
+    // Flag to indicate if the cache size needs to be reduced. This is set when the cache exceeds the maximum size.
+    private final AtomicBoolean cacheShrinkNeeded = new AtomicBoolean(false);
+    // Next timestamp to run reference queue purging to remove cleared references. Handled when cache is accessed.
+    private final AtomicLong nextReferenceQueuePurge = new AtomicLong();
+    // Deduplicates instances when the cached entry isn't in the actual cache.
+    // Holds weak references to the value so it won't prevent garbage collection.
+    private final Interner<V> valueInterner = Interners.newWeakInterner();
+
+    // Values are held as soft references to allow garbage collection when memory is low.
+    private final class SoftReferenceValue extends SoftReference<V> {
+        private final String key;
+
+        public SoftReferenceValue(String key, V referent, ReferenceQueue<? super V> q) {
+            super(referent, q);
+            this.key = key;
+        }
+
+        public String getKey() {
+            return key;
+        }
+    }
+
+    protected abstract V createValue(String key);
+
+    protected abstract int getCacheMaxSize();
+
+    protected abstract int getReduceSizeByPercentage();
+
+    protected abstract long getReferenceQueuePurgeIntervalNanos();
+
+    public void invalidateCache() {
+        cache.clear();
+    }
+
+    public V getIfPresent(String keyParam) {
+        SoftReferenceValue softReferenceValue = cache.get(keyParam);
+        return softReferenceValue != null ? softReferenceValue.get() : null;
+    }
+    public V get(String keyParam) {
+        // first do a quick lookup in the cache
+        V valueInstance = getIfPresent(keyParam);
+        if (valueInstance == null) {
+            // intern the topic name to deduplicate topic names used as keys, since this will reduce heap memory usage
+            keyParam = StringInterner.intern(keyParam);
+            // add new entry or replace the possible stale entry
+            valueInstance = cache.compute(keyParam, (key, existingRef) -> {
+                if (existingRef == null || existingRef.get() == null) {
+                    return createSoftReferenceValue(key);
+                }
+                return existingRef;
+            }).get();
+            if (cache.size() > getCacheMaxSize()) {
+                cacheShrinkNeeded.compareAndSet(false, true);
```

Comment:
```suggestion
                cacheShrinkNeeded.set(true);
```

No need for CAS here.